docs/middleware.rst

```diff
@@ -40,18 +40,78 @@ Pythonic
     where appropriate. For example, it converts the raw hex string returned by the RPC call
     ``eth_blockNumber`` into an ``int``.
 
-Built-in Middleware
+.. _Modifying_Middleware:
+
+Modifying Middleware
 ------------------
 
-Web3 ships with middleware for custom use, as desired. Middleware can be added after creating
-your Web3 object, like:
+Middleware can be added, removed, replaced, and cleared at runtime. To make that easier, you
+can name the middleware for later reference. Alternatively, you can use a reference to the
+middleware itself.
 
-.. code-block:: python
+.. py:method:: Web3.middleware_stack.add(middleware, name=None)
+
+    Middleware will be added to the top of the stack. That means the new middleware will modify the
+    request first, and the response last. You can optionally name it with any hashable object,
+    typically a string.
+
+    .. code-block:: python
+
+        >>> w3 = Web3(...)
+        >>> w3.middleware_stack.add(web3.middleware.pythonic_middleware)
+        # or
+        >>> w3.middleware_stack.add(web3.middleware.pythonic_middleware, 'pythonic')
+
+.. py:method:: Web3.middleware_stack.remove(middleware)
 
-    w3 = Web3(...)
-    w3.add_middleware(my_middleware)
+    Middleware will be removed from wherever it sat in the stack. If you added the middleware with
+    a name, use the name to remove it. If you added the middleware as an object, use the object to
+    remove it.
+
+    .. code-block:: python
+
+        >>> w3 = Web3(...)
+        >>> w3.middleware_stack.remove(web3.middleware.pythonic_middleware)
+        # or
+        >>> w3.middleware_stack.remove('pythonic')
+
+.. py:method:: Web3.middleware_stack.replace(old_middleware, new_middleware)
+
+    Middleware will be replaced wherever it sat in the stack. If the middleware was named, it will
+    continue to have the same name. If it was un-named, then you will now reference it with the new
+    middleware object.
+
+    .. code-block:: python
 
-Alternatively, you can pass in middlewares to the Web3 constructor.
+        >>> from web3.middleware import pythonic_middleware, attrdict_middleware
+        >>> w3 = Web3(...)
+
+        >>> w3.middleware_stack.replace(pythonic_middleware, attrdict_middleware)
+        # this is now referenced by the new middleware object, so to remove it:
+        >>> w3.middleware_stack.remove(attrdict_middleware)
+
+        # or, if it was named
+
+        >>> w3.middleware_stack.replace('pythonic', attrdict_middleware)
+        # this is still referenced by the original name, so to remove it:
+        >>> w3.middleware_stack.remove('pythonic')
```

Comment:
`default_normalization` seems reasonable (although maybe applies to returned vals better than request vals). Maybe `builtin_types` captures both directions? Either way, it probably belongs in a separate PR. Although conceptually simple, it will add a lot of line change noise to this already-long PR.